Support Updated api-version version

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

<!--- Please keep this note for the community --->

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->
Authentication using *`IDENTITY_ENDPOINT`* for Terraform in Azure Container App is using the the *`api-version`* *`2018-02-01`* as shown in the error below which is because of the default set in https://github.com/hashicorp/go-azure-sdk/blob/02fea5bf71c725cf01668bf28c523b7f84422718/sdk/auth/managed_identity_authorizer.go#L47.

```hcl
2023-05-01T13:05:58.920Z [WARN]  ValidateProviderConfig from "provider[\"registry.terraform.io/hashicorp/azurerm\"]" changed the config value, but that value is unused
2023-05-01T13:05:58.921Z [DEBUG] provider.terraform-provider-azurerm_v3.54.0_x5: Performing GET Request to "http://localhost:42356/msi/token?api-version=2018-02-01&client_id=003873c2-xxxx-xxxx-xxxx-xxxxxxxxxxxx&resource=https%3A%2F%2Fgraph.microsoft.com": timestamp=2023-05-01T13:05:58.921Z
2023-05-01T13:05:58.921Z [DEBUG] provider.terraform-provider-azurerm_v3.54.0_x5: GET http://localhost:42356/msi/token?api-version=2018-02-01&client_id=003873c2-xxxx-xxxx-xxxx-xxxxxxxxxxxx&resource=https%3A%2F%2Fgraph.microsoft.com: timestamp=2023-05-01T13:05:58.921Z
2023-05-01T13:05:58.924Z [DEBUG] provider.terraform-provider-azurerm_v3.54.0_x5: Reading Body from GET "http://localhost:42356/msi/token?api-version=2018-02-01&client_id=003873c2-xxxx-xxxx-xxxx-xxxxxxxxxxxx&resource=https%3A%2F%2Fgraph.microsoft.com": timestamp=2023-05-01T13:05:58.923Z
2023-05-01T13:05:58.924Z [ERROR] provider.terraform-provider-azurerm_v3.54.0_x5: Response contains error diagnostic: @caller=github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/internal/diag/diagnostics.go:55 @module=sdk.proto tf_provider_addr=provider tf_req_id=25930b13-aa7d-85a3-16d2-9f2dc2c91794 tf_rpc=Configure diagnostic_detail= diagnostic_severity=ERROR diagnostic_summary="building account: could not acquire access token to parse claims: ManagedIdentityAuthorizer: failed to request token from metadata endpoint: received HTTP status 400 with body: {"error":{"code":"UnsupportedApiVersion","message":"The HTTP resource that matches the request URI 'http://localhost:42356/msi/token' does not support the API version '2018-02-01'.","innerError":null}}" tf_proto_version=5.3 timestamp=2023-05-01T13:05:58.924Z
2023-05-01T13:05:58.924Z [ERROR] vertex "provider[\"registry.terraform.io/hashicorp/azurerm\"]" error: building account: could not acquire access token to parse claims: ManagedIdentityAuthorizer: failed to request token from metadata endpoint: received HTTP status 400 with body: {"error":{"code":"UnsupportedApiVersion","message":"The HTTP resource that matches the request URI 'http://localhost:42356/msi/token' does not support the API version '2018-02-01'.","innerError":null}}
2023-05-01T13:05:58.924Z [INFO]  backend/local: plan operation completed

Planning failed. Terraform encountered an error while generating this plan.

╷
│ Error: building account: could not acquire access token to parse claims: ManagedIdentityAuthorizer: failed to request token from metadata endpoint: received HTTP status 400 with body: {"error":{"code":"UnsupportedApiVersion","message":"The HTTP resource that matches the request URI 'http://localhost:42356/msi/token' does not support the API version '2018-02-01'.","innerError":null}}
│ 
│   with provider["registry.terraform.io/hashicorp/azurerm"],
│   on provider.tf line 1, in provider "azurerm":
│    1: provider "azurerm" {
│ 
╵
2023-05-01T13:05:58.940Z [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2023-05-01T13:05:58.943Z [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/hashicorp/azurerm/3.54.0/linux_amd64/terraform-provider-azurerm_v3.54.0_x5 pid=7848
2023-05-01T13:05:58.943Z [DEBUG] provider: plugin exited
```

The *`api-version`* for authenticating in Azure Container App is *`2019-08-01`*

### New or Affected Resource(s)/Data Source(s)

azurerm_resource_group

### Potential Terraform Configuration

```hcl
The api-version called when authenticating using IDENTITY_ENDPOINT will be 2019-08-01
```

### Service Name

Azure Container Apps

### API Versions Required

*`2019-08-01`*

### References

https://github.com/hashicorp/terraform-provider-azurerm/issues/21616

--- a/sdk/auth/auth.go
+++ b/sdk/auth/auth.go
@@ -133,9 +133,10 @@ func NewAuthorizerFromCredentials(ctx context.Context, c Credentials, api enviro
 
 	if c.EnableAuthenticatingUsingManagedIdentity {
 		opts := ManagedIdentityAuthorizerOptions{
-			Api:                           api,
-			ClientId:                      c.ClientID,
-			CustomManagedIdentityEndpoint: c.CustomManagedIdentityEndpoint,
+			Api:                             api,
+			ClientId:                        c.ClientID,
+			CustomManagedIdentityEndpoint:   c.CustomManagedIdentityEndpoint,
+			CustomManagedIdentityAPIVersion: c.CustomManagedIdentityAPIVersion,
 		}
 		a, err := NewManagedIdentityAuthorizer(ctx, opts)
 		if err != nil {

--- a/sdk/auth/config.go
+++ b/sdk/auth/config.go
@@ -7,7 +7,7 @@ import (
 	"github.com/hashicorp/go-azure-sdk/sdk/environments"
 )
 
-// Credentials sets up NewAuthorizer to return an Authorizer based on the provided credentails.
+// Credentials sets up NewAuthorizer to return an Authorizer based on the provided credentials.
 type Credentials struct {
 	// Specifies the national cloud environment to use
 	Environment environments.Environment
@@ -43,6 +43,9 @@ type Credentials struct {
 	EnableAuthenticatingUsingManagedIdentity bool
 	// CustomManagedIdentityEndpoint specifies a custom endpoint which should be used for Managed Identity.
 	CustomManagedIdentityEndpoint string
+
+	// CustomManagedIdentityAPIVersion specifies the API version to use for IMDS.
+	CustomManagedIdentityAPIVersion string
 
 	// Enables OIDC authentication (federated client credentials).
 	EnableAuthenticationUsingOIDC bool

--- a/sdk/auth/managed_identity_authorizer.go
+++ b/sdk/auth/managed_identity_authorizer.go
@@ -28,6 +28,10 @@ type ManagedIdentityAuthorizerOptions struct {
 	// CustomManagedIdentityEndpoint is an optional endpoint from which to obtain an access
 	// token. When blank, the default is used.
 	CustomManagedIdentityEndpoint string
+
+	// CustomManagedIdentityAPIVersion is an optional API version to use when requesting a token.
+	// This is required when using an endpoint that does not support the default API version such as Azure Container Apps.
+	CustomManagedIdentityAPIVersion string
 }
 
 // NewManagedIdentityAuthorizer returns an authorizer using a Managed Identity for authentication.
@@ -36,7 +40,7 @@ func NewManagedIdentityAuthorizer(ctx context.Context, options ManagedIdentityAu
 	if err != nil {
 		return nil, fmt.Errorf("determining resource for api %q: %+v", options.Api.Name(), err)
 	}
-	conf, err := newManagedIdentityConfig(*resource, options.ClientId, options.CustomManagedIdentityEndpoint)
+	conf, err := newManagedIdentityConfig(*resource, options.ClientId, options.CustomManagedIdentityEndpoint, options.CustomManagedIdentityAPIVersion)
 	if err != nil {
 		return nil, err
 	}
@@ -70,9 +74,9 @@ func (a *ManagedIdentityAuthorizer) Token(ctx context.Context, _ *http.Request) 
 		query["client_id"] = []string{a.conf.ClientID}
 	}
 
-	url := fmt.Sprintf("%s?%s", a.conf.MsiEndpoint, query.Encode())
+	u := fmt.Sprintf("%s?%s", a.conf.MsiEndpoint, query.Encode())
 
-	body, err := azureMetadata(ctx, url)
+	body, err := azureMetadata(ctx, u)
 	if err != nil {
 		return nil, fmt.Errorf("ManagedIdentityAuthorizer: failed to request token from metadata endpoint: %v", err)
 	}
@@ -135,16 +139,21 @@ type managedIdentityConfig struct {
 
 // newManagedIdentityConfig returns a new managedIdentityConfig with a configured metadata endpoint and resource.
 // clientId and objectId can be left blank when a single managed identity is available
-func newManagedIdentityConfig(resource, clientId, customManagedIdentityEndpoint string) (*managedIdentityConfig, error) {
+func newManagedIdentityConfig(resource, clientId, customManagedIdentityEndpoint string, customManagedIdentityAPIVersion string) (*managedIdentityConfig, error) {
 	endpoint := msiDefaultEndpoint
 	if customManagedIdentityEndpoint != "" {
 		endpoint = customManagedIdentityEndpoint
 	}
 
+	apiVersion := msiDefaultApiVersion
+	if customManagedIdentityAPIVersion != "" {
+		apiVersion = customManagedIdentityAPIVersion
+	}
+
 	return &managedIdentityConfig{
 		ClientID:      clientId,
 		Resource:      resource,
-		MsiApiVersion: msiDefaultApiVersion,
+		MsiApiVersion: apiVersion,
 		MsiEndpoint:   endpoint,
 	}, nil
 }

